From 0e786f459d274cf9a8afc87221bbb1c77ccd8eae Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Miha=20=C4=8Can=C4=8Dula?= <miha@noughmad.eu>
Date: Thu, 26 May 2011 22:07:49 +0200
Subject: [PATCH] Load the correct default theme when using KGR

---
 src/core/renderer.cpp |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/core/renderer.cpp b/src/core/renderer.cpp
index 992b103..14e688b 100644
--- a/src/core/renderer.cpp
+++ b/src/core/renderer.cpp
@@ -25,9 +25,9 @@ using namespace Knights;
 
 #if defined WITH_KGR 
 
-Renderer::Renderer ( const QString& defaultTheme ) : KGameRenderer ( defaultTheme )
+Renderer::Renderer ( const QString& defaultTheme ) : KGameRenderer ( QLatin1String("themes/default.desktop") )
 {
-
+    setTheme ( defaultTheme );
 }
 Renderer::~Renderer () 
 {
-- 
1.7.5.1

